### ISSUE TITLE
'mvn site' fails for FindBugs report, causing 'site' goal to fail

### ISSUE THREAD

**Issue by kwwall** [ESAPI/esapi-java-legacy#387]
Maven produces the following error message on ESAPI's latest pom.xml (from the commit made on 4/16/2017):
`[INFO] All plugins with a version specified are using the latest versions.

[INFO]
[INFO] All plugins have a version specified.
[INFO]
[INFO] Project defines minimum Maven version as: 3.0
[INFO] Plugins require minimum Maven version of: 3.1
[INFO]
[ERROR] Project requires an incorrect minimum version of Maven.
[ERROR] Either change plugin versions to those compatible with 3.0
[ERROR] or update the pom.xml to contain
[ERROR]     &lt;prerequisites&gt;
[ERROR]       &nbsp;&nbsp;&nbsp;&nbsp;&lt;maven&gt;3.1&lt;/maven&gt;
[ERROR]     &lt;/prerequisites&gt;
[INFO]
[INFO] Require Maven 3.0.1 to use the following plugin updates:
[INFO]   org.codehaus.mojo:findbugs-maven-plugin ....................... 3.0.3
[INFO]`

I think there are at least 2 ways to address this. One is to require Maven version 3.1 or later. The other is to find a previous `org.codehaus.mojo:findbugs-maven-plugin` version that works with Maven 3.0. (The version that currently is specified is 2.5.5.)

I am somewhat reluctant to require Maven 3.1 as Maven 3.0 can even be a minor PITA to install (at least on Linux systems where the current deb or rpm package wants to pull in some ancient Maven 2.x version instead). E.g., on Linux Mint 17.2, an LTS version based on Ubuntu 14.04.2 LTS, I had to manually install Maven 3.0.5 to get it to work. I'm not sure everyone who wants to build ESAPI is up for that level of masochism, so I'd like to discuss the options.  We could keep the Maven 3.0 requirement instead of upping it to 3.1 if we can find out what version of FindBugs maven plugin works with Maven 3.0.

Let's discuss people's preferences!
-kevin

**Comment by augustd** [comment]
3.1 was released in July 2013, so I don't think it is that onerous to expect developers to have a newer version than that! It was fairly simple to update my IDE to point to a newer version of maven (v3.3.9) and then update pom.xml to require 3.1: 

```
    <prerequisites>
        <maven>3.1</maven>
    </prerequisites>
```

I can send a pull request if you like...

**Comment by augustd** [comment]
Now that the project is compiling with maven 3.1, OWASP Dependency Check will actually run, and fail the build: 

```
Failed to execute goal org.owasp:dependency-check-maven:1.4.4:check (default) on project esapi:

Dependency-Check Failure:
One or more dependencies were identified with vulnerabilities that have a CVSS score greater then '1.0':
commons-fileupload-1.3.2.jar: CVE-2016-1000031
commons-httpclient-3.1.jar: CVE-2015-5262, CVE-2014-3577, CVE-2012-6153
servlet-api-2.5.jar: CVE-2006-5654, CVE-2006-3225, CVE-2006-2501
See the dependency-check report for more details.
-> [Help 1]
```

**Comment by kwwall** [comment]
I knew that I broke "mvn site" with the last commit to 'pom.xml'...I think
I even left a note to that effect in the commit comment, although it may
have just been in a private email to Katy Aton.

I was reluctant to mandate an update to Maven because some people are using
various Linux LTS releases that still don't even support Maven 3.0 and thus
have to be updated manually l and that is somewhat of a PITA. But OTOH, I
didn't realize that the 3.1 release was so old either. So, no pain, no gain
I guess.

-kevin
P.S.- If the pom.xml changes are more extensive that the 1 line change to
the Maven release #, then a PR would be much appreciated. But even I should
be able to handle a single line change! :)
--
Blog: http://off-the-wall-security.blogspot.com/  |  Twitter:  @KevinWWall
NSA: All your crypto bit are belong to us.

On May 2, 2017 5:13 PM, "August Detlefsen" <notifications@github.com> wrote:

3.1 was released in July 2013, so I don't think it is that onerous to
expect developers to have a newer version than that! It was fairly simple
to update my IDE to point to a newer version of maven (v3.3.9) and then
update pom.xml to require 3.1:

    <prerequisites>
        <maven>3.1</maven>
    </prerequisites>

I can send a pull request if you like...

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/ESAPI/esapi-java-legacy/issues/387#issuecomment-298762940>,
or mute the thread

**Comment by kwwall** [comment]
Dang. It's been a long while (years?) since I ran Dependency Check. I
probably should bump the CVSS2 figure up a bit. I had it set to 1 mostly
for testing. Will fix the dependencies at that time as well.

-kevin
--
Blog: http://off-the-wall-security.blogspot.com/  |  Twitter:  @KevinWWall
NSA: All your crypto bit are belong to us.

**Comment by kwwall** [comment]
Well, this isn't going to be so simple. I updated to Maven 3.1.1 and updated the pom.xml to make it require maven 3.1 as @augustd  mentioned above. I was hoping that that fixed the issue with 'mvn site' failing with the FindBugs report, but it didn't.

Then I looked at the output of the Dependency Check report and researched those a bit. Where to start.

Okay, first, I had set `<failBuildOnCVSS>1</failBuildOnCVSS>` so that it would pretty much fail if _any_ dependency had a reported CVE. That was only intended for testing and I was going to bump it up to something more reasonable, like maybe 4 or 5. Well, crap! The first dependency that Dependency Check reports as vulnerable is CVE-2016-1000031 reported for commons-fileupload-1.3.2.jar. What sucks is that has a CVSSv3 score of 9.8 and a CVSSv2 score of 7.5, so even raising the 'failBuildOnCVSS' level to something reasonable like 5 is not going to cause Dependency Check to pass. However, a bigger problem is that releawse 1.3.2 is the latest official Apache release for Apache Commons FileUpload. Apache has not "fixed" this because they contest that it is not actually a problem with their library.

The good news is that my research turned up this [Tenable link](http://www.tenable.com/security/research/tra-2016-12) that states that this is only a security issue if an application intends to deserialize DiskFileItems, but since ESAPI does not use this class at all, ESAPI is not vulnerable to this CVE. (ESAPI uses DiskFileItemFactory, but not DiskFileItems.) I believe that there is a way to quiet Dependency Check regarding specific CVEs; short of that, the only way to get rid of this warning would be to get rid of the Apache Commons FileUpload dependency completely.

However, there is some more bad news...not insurmountable, but in order to address the 3 CVEs associated with servlet-api-2.5.jar, I think that we will need to upgrade to release 3.0.1 or 3.1.0, neither of which is in the Maven Central repository. I am reluctant to start pulling various jars from several repos though, so if anyone has any suggestion on an alternate approach, I'd love to hear it.

**Comment by augustd** [comment]
I've just sent PR #388 which shows how to suppress the dependency check plugin from failing the build due to CVE-2016-1000031.

**Comment by kwwall** [comment]
LOL. Thanks, I just researched and found that today, but this will save me
some time.

-kevin
--
Blog: http://off-the-wall-security.blogspot.com/  |  Twitter:  @KevinWWall
NSA: All your crypto bit are belong to us.

**Comment by augustd** [comment]
@kwwall I've updated PR #388 so that it fully compiles, all tests pass, and no CVSS > 5.8 in dependencies. 

I updated to servlet API 3.0.1 (released July 2011). It _is_ in Central, but the artifact ID changed to `javax.servlet-api`. This necessitated some changes to the tests, to implement new methods in servlet API interfaces. An alternative might be to use a mocking framework like Mockito, but for now I just wanted a version that compiles.

**Comment by augustd** [comment]
Should this be closed now that PR #388 is merged?

**Comment by kwwall** [comment]
No, because 'mvn site' is *still* failing the findbugs report generation,
at least for me. Does it work for you? If so, it may be an environmental
issue on my side.

-kevin
--
Blog: http://off-the-wall-security.blogspot.com/  |  Twitter:  @KevinWWall
NSA: All your crypto bit are belong to us.

On May 5, 2017 3:19 AM, "August Detlefsen" <notifications@github.com> wrote:

@kwwall <https://github.com/kwwall> I've updated PR #388
<https://github.com/ESAPI/esapi-java-legacy/pull/388> so that it fully
compiles, all tests pass, and no CVSS > 5.8 in dependencies.

I updated to servlet API 3.0.1 (released July 2011). It *is* in Central,
but the artifact ID changed to javax.servlet-api. This necessitated some
changes to the tests, to implement new methods in servlet API interfaces.
An alternative might be to use a mocking framework like Mockito, but for
now I just wanted a version that compiles.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/ESAPI/esapi-java-legacy/issues/387#issuecomment-299398828>,
or mute the thread
